### DISCUSSION TITLE
Windows 10 - 11 installation on a iSCSI target

### DISCUSSION THREAD

**Discussion by sebaxakerhtc** [ipxe/ipxe#324] [Show and tell]
As many of us knows - after 1st step of installation (copyng files) we can't boot from target because of BSOD's
Do not forget to add `keep` key to yor sanboot command

1. Solution (E1000e specific): If you have BSOD `e1i63x64.sys is not digitaly signed` or `INACCESSIBLE_BOOT_DEVICE` for E1000e ethernet controller. 
You need to create a new WinPE with ADK 2004. This driver will be replaced with e1i65x64.sys
2. [Solution from forum (Main)](https://forum.ipxe.org/showthread.php?tid=23709&pid=45846#pid45846): `DRIVER_IRQL_NOT_LESS_OR_EQUAL` You need to edit registry after 1st step of installation. Attach your iSCSI drive to another OS and remove data from \SYSTEM\ControlSet001\Control\Session Manager\Memory Management\PagingFiles
See screenshots below
![1](https://user-images.githubusercontent.com/32651506/114939540-41487180-9e49-11eb-90e6-48720950c74c.png)
![2](https://user-images.githubusercontent.com/32651506/114939542-43123500-9e49-11eb-97fa-9a9086339447.png)

**Comment by PatcoreShare** [comment]
Hey, I have a problem with using this method.
Windows 10 1803 works (no tricks), 
Win 10 2004 don't works 
![wersja windows](https://user-images.githubusercontent.com/11398415/128996062-8c9d0f81-87a4-4904-afa0-a86e4fdaea52.png)

Prepare WinPE
![image](https://user-images.githubusercontent.com/11398415/128996156-77618c42-c110-4d52-a4d0-5ff3e886cce5.png)
![image](https://user-images.githubusercontent.com/11398415/128996182-aac228c7-6cac-4383-a910-af54efc97928.png)
![image](https://user-images.githubusercontent.com/11398415/128996190-962c0252-32c5-45d3-83a9-2964b91566ca.png)
![image](https://user-images.githubusercontent.com/11398415/128996202-650bb40f-a416-4029-b251-f68d298eee56.png)


Configuration iPXE  
![bootipxe](https://user-images.githubusercontent.com/11398415/128996291-125637c6-ce38-42f3-be71-017dc3740cc2.png)

Start installation.

![winpesetup](https://user-images.githubusercontent.com/11398415/128996343-ba7747e9-758d-4bef-8a09-2d33dce86df3.png)

Choosing a Windows.
![selectWin](https://user-images.githubusercontent.com/11398415/128996379-dc4c7c23-383b-4e75-9b16-4efc672a3bb2.png)


First stage is over.

On a different computer
![iscsi](https://user-images.githubusercontent.com/11398415/128996443-e69276da-25f0-4ea8-abf1-301b4dbc6b92.png)

![importregedit](https://user-images.githubusercontent.com/11398415/128996449-69ea669a-c7f2-47a2-8226-e97f3f62589d.png)


![iscsinieaktywne](https://user-images.githubusercontent.com/11398415/128996484-ea8cddeb-e023-4fc7-8e85-436eb511fac0.png)

change configuration 

![bootipxe-boot](https://user-images.githubusercontent.com/11398415/128996532-956aaf63-6695-431e-9067-bd3ef794e078.png)

trouble

![bladpoinstalacji](https://user-images.githubusercontent.com/11398415/128996559-42290d51-b396-40ff-9aa7-29754f264f33.png)

I try F8 and F7  (it used to work)

Any suggestions?

**Reply by sebaxakerhtc** [reply]
> I am trying on `Windows 10 21H1` and `Windows Server 2019`. It seems like setting `PagingFiles` to empty string isn't necessary. I can't boot to the Windows from iSCSI smoothly.

Look at solution 1 from my post.
You need to update your WinPE because of e1i63x64.sys

**Reply by PatcoreShare** [reply]
How I can update WinPE? 
Some "how to" for someone new

**Reply by sebaxakerhtc** [reply]
> How I can update WinPE?
> Some "how to" for someone new

Install [ADK + WinPE add-on](https://docs.microsoft.com/en-us/windows-hardware/get-started/adk-install) and read [instructions](https://docs.microsoft.com/en-us/windows-hardware/manufacture/desktop/winpe-create-usb-bootable-drive)
Don't forget to add iSCSI support to WinPE (WinPE-StorageWMI). Use google search on how-to's

**Reply by ccy** [reply]
I am using Intel NUC 10th Gen. The network card is Intel I219-V.  Unfortunately, intel don't supply driver for Windows Server 2019.  I am using the Windows ADK 19041 to prepare the WinPE

**Reply by sebaxakerhtc** [reply]
> I am using Intel NUC 10th Gen. The network card is Intel I219-V. Unfortunately, intel don't supply driver for Windows Server 2019. I am using the Windows ADK 19041 to prepare the WinPE

I can't help you with it. My solution is for the E1000e LAN card.

**Reply by sebaxakerhtc** [reply]
> The network card is Intel I219-V

Did you try to add drivers to the WinPE and Windows image using DISM?
[Like this](https://docs.microsoft.com/en-us/windows-hardware/manufacture/desktop/add-and-remove-drivers-to-an-offline-windows-image)
I successfuly added drivers for the intel 11th gen drives (intel VMD, iRST) to my images

**Reply by PatcoreShare** [reply]
Unfortunately, I still have a problem and I am asking for help.

What has been done:
1.
Install ADK and ADK Addon  (https://docs.microsoft.com/en-us/windows-hardware/get-started/adk-install)
![image](https://user-images.githubusercontent.com/11398415/129198200-5dd4a95b-602b-415f-be91-5aa91f78fcde.png)
2. In Deployment and Imaging Tools Environment
`copype amd64 D:\WinPE`
3. Mount image
`Dism /mount-image /imagefile:”d:\winPE\media\sources\boot.wim” /index:1 /mountdir:”d:\winTemp”`
4. https://docs.microsoft.com/en-us/windows-hardware/manufacture/desktop/winpe-add-packages--optional-components-reference) 
![image](https://user-images.githubusercontent.com/11398415/129198364-5f602ac5-9236-4da6-95b6-b384f6f34a82.png)
`Dism /Add-Package /Image:"d:\WinTemp" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-WMI.cab"`

`Dism /Add-Package /Image:"d:\WinTemp" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\en-us\WinPE-WMI_en-us.cab"`

`Dism /Add-Package /Image:"d:\WinTemp" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-NetFx.cab"`

`Dism /Add-Package /Image:"d:\WinTemp" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\en-us\WinPE-NetFx_en-us.cab"`

`Dism /Add-Package /Image:"d:\WinTemp" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-Scripting.cab"`

`Dism /Add-Package /Image:"d:\WinTemp" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\en-us\WinPE-Scripting_en-us.cab"`

`Dism /Add-Package /Image:"d:\WinTemp" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-PowerShell.cab"`

`Dism /Add-Package /Image:"d:\WinTemp" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\en-us\WinPE-PowerShell_en-us.cab"`

`Dism /Add-Package /Image:"d:\WinTemp" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\WinPE-StorageWMI.cab"`

`Dism /Add-Package /Image:"d:\WinTemp" /PackagePath:"C:\Program Files (x86)\Windows Kits\10\Assessment and Deployment Kit\Windows Preinstallation Environment\amd64\WinPE_OCs\en-us\WinPE-StorageWMI_en-us.cab"`


5. Commit
```
dism /commit-image /mountdir:"d\winTemp"
dism /unmount-image /mountdir:"d:\WinTemp" /discard
```

Not working, still `e1i63x64.sys is not digitaly signed`

I download this: 
![image](https://user-images.githubusercontent.com/11398415/129199035-e4805b04-cd2f-4071-a92c-1092c3f71c5c.png)

I got this result:
![image](https://user-images.githubusercontent.com/11398415/129199113-147f88a2-08dc-4f64-b6d9-0f7e2114aa0e.png)

Where did I go wrong, help me, please.

**Comment by sebaxakerhtc** [comment]
Just NOW i see you are use already 22000.1 from Windows 11., it buggy. For example i can't add Russian language - it gives me error... Very strange.
May be there is a problem. Try to download and install older one ADK
![Снимок экрана 2021-08-11 в 17 45 41](https://user-images.githubusercontent.com/32651506/129050763-7f6c6b0d-65a2-4ceb-a9a5-e76991c8fab1.png)

**Comment by sebaxakerhtc** [comment]
First:
Are you use E1000e or other network card?
Second:
Use this command when unmount instead of your 2 commands

dism /unmount-image /mountdir:"d:\WinTemp" /commit


чт, 12 авг. 2021 г., 15:48 Patryk ***@***.***>:

**Reply by PatcoreShare** [reply]
> Are you use E1000e or other network card?

I use VMware for testing. 
> 2. Solution from forum (Main) with PagingFiles

It doesn't work, I'm looking for a solution

**Comment by sebaxakerhtc** [comment]
And I use VMWare and all works, but you should set E1000e for the LAN card
in settings of VM

чт, 12 авг. 2021 г., 16:10 Patryk ***@***.***>:

**Comment by oschroeder8** [comment]
I wrote a script as part of some other work I was doing:
https://github.com/oschroeder8/provision-content/tree/preliminary-untested-winpe-support/tools/windows

The .CMD script can be used to build a WinPE with any Windows 10/11 ADK with WinPE add-on.

It also provides for a folder where you can put driver packages, in your case you could extract the Intel NIC driver package, and if you wanted to, even the VMware VMXNET3 driver. That way your WinPE would have those drivers included.

**Comment by sebaxakerhtc** [comment]
Windows 11 installed on a iSCSI disk (tgt)
![Screenshot 2021-10-12 143336](https://user-images.githubusercontent.com/32651506/137028208-4391959c-5b23-47e3-bc03-159f166be0a0.png)

**Reply by gh-crawler** [reply]
Hi. Do you have an expirience in moving installed Windows from local drive to iSCSI? I fixed BCD (To avoid of 0x0E error: winload.exe not found) and now get BSOD during Windows splash screen: INACCESSIBLE BOOT DEVICE. The lab is on VMware. I've already tried to swithc NIC from E1000 to VMXNET 3 - doesn't work. Tried to set net0/gateway 0.0.0.0 - doesn't work

**Reply by stappersg** [reply]
Time will tell when the above comment gets marked with _This comment was marked as off-topic._

https://github.com/ipxe/ipxe/pull/370#issuecomment-1484045058 got such mark within a hour.

Message of that comment was and is:  _"Merge request #370 deserves attention"_

Message of this comment: _"I have the slight feeling that "those who can" use (abuse?) their power to silence me"_

What I think would be good for the  iPXE  project:

* Less dangling open merge requests  (e.g. #370)
* Ignore what can be ignored
* Discuss what should be discussed

**Reply by NiKiZe** [reply]
@stappersg this will only be said once. Stop creating useless, non constructive comments which just steals focus from more important tasks.

Even if your goal is good, these comments are not acceptable.

**Comment by sebaxakerhtc** [comment]
Windows 11 22H2 installed on iSCSI storage on intel NUC10 Performance Kit with l219-v (targetcli-fb)
![screenshot](https://user-images.githubusercontent.com/32651506/192783097-edf55143-d33e-4c95-a300-ac34499609c4.png)

**Comment by firemeteorxx** [comment]
Hello, people in this thread. I'm writing to query if anybody has the experience of installing Windows 10 22H2 or upgrading from an older win 10 version to it. I'm running into problem for both path and would like to understand if this is an issue specific to my situation or a common issue.

I'm able to get Win10 1709 / 1809 booting from iSCSI disk. But both of them are fading away...

**Reply by stappersg** [reply]
@firemeteorxx could have add a link to https://github.com/ipxe/ipxe/discussions/817 with an invite to continue over there ...

**Reply by firemeteorxx** [reply]
Thank you for providing the link. Any information / discussion will be appreciated.

The successful story about Win11 in this thread is encouraging. But I'm not sure if my old computer is qualified for Win 11, or if I can upgrade to Win 11 with my existing Win10 license? That's why I haven't looked into this possibility.

**Comment by epacua** [comment]
I have.  In addition, I have win7, winxp and various linux distros.

![image](https://user-images.githubusercontent.com/15176373/213339298-f72f57b6-963e-4e1a-be14-34bf7f770cb2.png)

**Comment by antermin** [comment]
I ran into a weird issue - I can `sanhook` in iPXE and install Windows 10 (21H2) to the iSCSI target, but I can't boot it using `sanboot` unless I use the iSCSI initiator in my NUC's BIOS.

Background:
Legacy boot is disabled. (Tried booting WinPE in legacy mode but it didn't work, and also the BIOS' iSCSI initiator is only available in UEFI mode.)

My steps:
1. `sanhook` iSCSI target, `wimboot` into WinPE files and install Windows 10.
2.  If I attempt to use iPXE's `sanboot` for stage 2 installation, it shows "The operating system couldn't be loaded because the kernel is missing or contains errors." "File: \Windows\system32\ntoskrnl.exe".
3. Switch to BIOS' iSCSI initiator, and the installation is successful. Afterwards I can boot into the OS if I use BIOS' iSCSI initiator.
4. Switch to iPXE's `sanboot` and it shows "Preparing Automatic Repair", and can only boot into the Windows recovery environment.

I tried `keep-san` / `--keep`, setting / unsetting gateway to `0.0.0.0`, `--no-describe`, etc. but it still didn't work.

I also tried using another PC (non-NUC) to boot but it also results in "Preparing Automatic Repair".

**Comment by epacua** [comment]
`set net0/gateway 0.0.0.0` is pretty important. I suspect it's related to initiator-iqn issue. Does your target contains multiple luns? If so, you would need to specify the initiator-iqn to connect to which lun. For example, as per my config below, if i wish to connect to win10, my ipxe menu should be like this:

`96 :win10_x64-iscsi  
97     echo Booting Windows 10 over iscsi  
98     set net0/gateway 0.0.0.0
99     set initiator-iqn iqn.2021-10.yyy.xxxx.internal.${hostname}:win10
100     sanboot iscsi:${server}::::iqn.2021-10.yyy.xxxx.internal.${hostname}:os || goto failed  

`


![image](https://user-images.githubusercontent.com/15176373/215367677-e15c8f64-0d95-4438-8d40-15d9b1442058.png)

**Reply by antermin** [reply]
My target has only 1 LUN, NUC only has 1 NIC, and my script looks like the following:
```
set gateway 0.0.0.0
set initiator-iqn iqn.2000-01.com.xxx:windows
sanboot --keep iscsi:${server}:::0:iqn.2000-01.com.xxx:win10
```
Hooking works with the same initiator-iqn and target string, but booting it does not work.
And when I input the same strings to my NUC BIOS' iSCSI configuration, I can select "Windows Boot Manager" and network boot Windows 10.

**Reply by epacua** [reply]
Here are the steps that I would try:
1- Remove `--keep`, that is not needed.
2- Drop to a shell if sanboot fails, eg `sanboot ... || shell` so I can inspect the variables, ping the target etc
3- Confirm that both the boot and install.wim contain the iscsi objects. See discussion [here](https://github.com/ipxe/ipxe/discussions/817#discussioncomment-4775131).
4- Confirm that nothing is accessing the lun by accident. (In my other setup, I use separate user:password for each target to ensure that nothing can log into it by accident).
5- Try other machine or vm
6- Try ipxe.pxe

By the way, what's your iscsi target? Here's my wimboot block
`
#!ipxe
kernel http://${server}/pxeboot/winpe/wimboot
initrd http://${server}/pxeboot/winpe/${arch}/winpe10/boot/bcd                     bcd
initrd http://${server}/pxeboot/winpe/${arch}/winpe10/boot/boot.sdi                boot.sdi
initrd http://${server}/pxeboot/winpe/${arch}/winpe10/sources/boot.wim             boot.wim
boot  

`

**Reply by antermin** [reply]
So I stumbled upon @Dids' [comment](https://forums.unraid.net/topic/98483-plugins-iscsi-gui-and-iscsi-target/page/4/#comment-1014821) because I also found the same error message in `dmesg` as in [this issue](https://github.com/SimonFair/unraid.iSCSI/issues/4). I am using [targetcli-fb](https://aur.archlinux.org/packages/targetcli-fb/) on Arch Linux.

Once I switched from fileio (the file was placed on an ext4 partition) to a block device (a partition dedicated for the iSCSI target), I can use iPXE to boot Windows 10 over the network.

**Comment by Fibbicles** [comment]
Anyone know why I might get "Windows cannot be installed to this disk. This computer's hardware may not support booting to this disk. Ensure that the disk's controller is enabled in the computer's BIOS menu." when in Legacy BIOS and works fine in UEFI?

**Reply by sebaxakerhtc** [reply]
Place LAN boot at the first place in boot order in BIOS settings

**Comment by adrian-agp** [comment]
Hello, what could be happening that when selecting the iscsi disk to install it remains at 0 copying Windows files (0%)

**Reply by Fibbicles** [reply]
How have you got your targets configured? Are they using writethrough/writeback mode?

**Comment by windisklessfree** [comment]
Windows bsod by iscsi.
arises from the network infrastructure, namely the competitive dependence between each other NDIS and Netio. Here is a tool that helps solve this problem in 5 minutes. This always helps me. https://github.com/windisklessfree/IRQ_NOT_LESS_OR_EQUAL Plus optimizer working iscsi
I hope this helps) Thank you!
